### ISSUE TITLE
Speeding up dataset and dataloader

### ISSUE THREAD

**Issue by frostedoyster** [lab-cosmo/metatensor#482]
I'm wondering if it's possible to torchscript the indexing and collating in the Datasets. It made a huge difference to me: from super-slow training to relatively fast
@jwa7 @agoscinski @abmazitov @Luthaf

**Comment by jwa7** [comment]
Sounds good! Do you have example code that is half ready to go? If so we can start there and iterate together :)

**Comment by frostedoyster** [comment]
I only have the code in metatensor-models.
https://github.com/lab-cosmo/metatensor-models/blob/df96bb062ddf9e7ee2ff6ed620f35e788487ec6d/src/metatensor/models/utils/data/dataset.py#L20-L21
We could aim to just compile these functions. I fear compiling the whole index functions within the Datasets and the collate functions would be much harder

**Comment by ceriottm** [comment]
Can you profile to figure out what is making the current implementation that slow?

**Comment by frostedoyster** [comment]
Completely in RAM

**Comment by frostedoyster** [comment]
https://gist.github.com/frostedoyster/eb82bdf463c05cfd04cb6b652940d624
https://gist.github.com/frostedoyster/27a1d45eb350e0baf51e62d34dd26463

With these simple examples, I get 5x on `join` and 30x on `slice` by compiling. I'll try if I can profile

**Comment by frostedoyster** [comment]
I don't know if join can be sped up without compiling. Regarding slice,

```
all_samples = block.samples.view(labels.names).to_owned()
# create an arrays of bools indicating which samples indices to keep
_, position, _ = all_samples.intersection_and_mapping(labels)
samples_mask = _dispatch.where(position != -1, True, False)
```

This would work in principle to substitute the list above, but `to_owned()` fails because some entries can be repeated in the view, and owned labels can't have the same entry multiple times. For now, I think the easiest option is to compile

**Comment by Luthaf** [comment]
I'm surprised by how much of a different torch script makes. From my understanding, it should mainly be replacing the overhead of running the Python, interpreter with the overhead of running the TorchScript interpreter. Even if TorchScript is a lot simpler than Python, that's still quite a large difference!

> This would work in principle to substitute the slow list above, but to_owned() fails because some entries can be repeated in the view, and owned labels can't have the same entry multiple times. For now, I think the easiest option is to compile

We could make it so the right hand side of `intersection_and_mapping` does not need to be owned, only the left hand  side one. But this might require fairly intrusive changes so I'd go for compiling anyway in the mean time!

**Comment by jwa7** [comment]
It sounds like this is more an issue with the speed of `join` and `slice`, with downstream implications on specific collate functions, as opposed to an issue with the `Dataset` and `DataLoader`. We could, for instance, provide options to jit compile collate functions when instantiating the `DataLoader`. For example, the default collate `group_and_join` could use a compiled `join`, given the speed up shown by Filippo.

Is letting the user fine-control when and where compilation happens something we want to provide in general? Or is this over-engineering?

**Comment by Luthaf** [comment]
Doing compilation by default sounds like a good idea. We might also want to look into speeding these functions up, but compilation should provide a fix for now. 

The only issue I foresee is that we can only compile with types from metatensor-torch, so only the version available as `metatensor.torch.learn.Dataset` would use the compiled code.

**Comment by Luthaf** [comment]
I'm going to close this issue for now, if more slowness is found in a specific part of the code we can open a new one!
